src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ThreadPoolValueTaskSource.cs

```diff
@@ -0,0 +1,187 @@
+        /// be completed synchronously on the thread pool.
+        /// </summary>
+        internal sealed class ThreadPoolValueTaskSource : IThreadPoolWorkItem, IValueTaskSource<int>, IValueTaskSource<long>
+        {
+            private enum Operation : byte
+            {
+                None,
+                Read,
+                Write,
+                ReadScatter,
+                WriteGather
+            }
+
+            private ManualResetValueTaskSourceCore<long> _source;
+            private readonly SafeFileHandle _fileHandle;
+            private Operation _operation = Operation.None;
+
+            // These fields store the parameters for the operation.
+            // The first two are common for all kinds of operations.
+            private long _fileOffset;
+            private CancellationToken _cancellationToken;
+            // Used by simple reads and writes. Will be unsafely cast to a memory when performing a read.
+            private ReadOnlyMemory<byte> _singleSegment;
+            // Used by vectored reads and writes. Is an IReadOnlyList of either Memory or ReadOnlyMemory of bytes.
+            private object? _multiSegmentCollection;
+
+            internal ThreadPoolValueTaskSource(SafeFileHandle fileHandle)
+            {
+                _fileHandle = fileHandle;
+                _source.RunContinuationsAsynchronously = true;
+            }
+
+            private long GetResultAndRelease(short token)
+            {
+                try
+                {
+                    return _source.GetResult(token);
+                } finally
+                {
+                    _source.Reset();
+                    _fileHandle.TryToReuse(this);
+                }
+            }
+
+            public ValueTaskSourceStatus GetStatus(short token) => _source.GetStatus(token);
+            public void OnCompleted(Action<object?> continuation, object? state, short token, ValueTaskSourceOnCompletedFlags flags) =>
+                _source.OnCompleted(continuation, state, token, flags);
+            int IValueTaskSource<int>.GetResult(short token) => (int) GetResultAndRelease(token);
+            long IValueTaskSource<long>.GetResult(short token) => GetResultAndRelease(token);
+
+            void IThreadPoolWorkItem.Execute()
+            {
+                Debug.Assert(_operation >= Operation.Read && _operation <= Operation.WriteGather);
+
+                long result = 0;
+                try {
+                    bool notifyWhenUnblocked = false;
+                    try
+                    {
+                        // This is the operation's last chance to be cancelled.
+                        if (_cancellationToken.IsCancellationRequested)
+                        {
+                            _source.SetException(new OperationCanceledException(_cancellationToken));
+                            return;
+                        }
+
+                        notifyWhenUnblocked = ThreadPool.NotifyThreadBlocked();
```

Comment:
I thought to take advantage of #53471 to maybe help scenarios with lots of thread pool threads blocked due to I/O.

Ideally this should run on the thread pool's I/O threads but there is no API to schedule an `IThreadPoolWorkItem` there, and they are still written in unmanaged code.